### ISSUE TITLE
Smart Object

### ISSUE THREAD

**Issue by milqnomadu** [EmilDohne/PhotoshopAPI#90]
(1) I was trying to replace the image inside a smart object, save it and then export the psd as an image. Using psd-tools this isn't possible and the author said it is not planned (although it is massively wanted).

(2) Then tried instead to retrieve the information from the smart object and have it replicated in PIL: opacity, position, bbox, perspective and warp. However using psd-tools Warp I'm not able to retrieve.

Is it possible to either (1) or (2) with this repo?

Sorry if I disturbed you in any way, hope you have a great day!

**Comment by EmilDohne** [comment]
Hey @milqnomadu thanks for checking this repo out!

While its not possible currently it is planned to support all of these features although some are more involved than others.

This problem can be broken down into 3 separate problems I suppose:

- Parsing of Smart Objects (planned with the 0.5.0 release)
- A rendering pipeline that allows for rendering of arbitrary layers (text, smart object etc.) As well as for composing layers with blend modes (planned somewhere before the 1.0 release)
- allowing for additional transforms such as warps etc.

I have recently been occupied with another project which I plan on completing within the next month allowing me to focus more on these features.

Im not sure how well versed you are in C++ but if you have some time to spare contributions are always welcome in any of these departments.

Hopefully this answers your question, unfortunately I dont think theres currently a project which has these features implemented but I am very interested in pursuing them :)

**Comment by milqnomadu** [comment]
Hey Emil, thank you so much for the response! I love your work and I'm so thriving for your contribution. 

Unfortunately I'm incredibly ignorant on the subject as I'm 19yrs old from Italy and I'm just starting to explore these projects on my own. I just need something that lets me manually set up and transform a shape that will then be used as mask to overlay a bunch of other images for a urgent project of mine, so either replacing the image inside a SO and then export the psd as webp automatically or (in order to not reinvent the wheel) just to retrieve the necessary information and then apply them with PIL. 

I also tried to just replicate a sort of smart object in Python with GUI (getting help from claude 3.5) where I managed to achieve all the modifications I needed, however having troubles with replicating the warp and also researching on it's theory and having being banned from stackoverflow for asking a question regarding the warp (absurd) I had to stop.

I will be ecstatically looking forward for the updates and I'm 150% routing for you. Did you have a timeline for the SO update?
If you want moral support we can definitely connect and I'll provide for sure ahaha, unfortunately can't provide much code knowledge, you're way smarter than me.

**Comment by EmilDohne** [comment]
Hey @milqnomadu, unfortunately I am unable to give an exact timeline as from researching smart objects earlier they are not necessarily trivial to implement. But I am guessing somewhere between Q3 and Q4 2024 it should be possible :)

**Comment by milqnomadu** [comment]
Hey Emil thank you so much! Will wait anxiously for the update that will enable to replace images inside of smart objects. I'm sure you can do it even faster than the predicted time. If you need any type of help I will be very happy to help. 

Have an awsome day!

**Comment by EmilDohne** [comment]
@milqnomadu to give you an update on the progress you can follow #57 as i've started working on this issue and will update the status in there :)

**Comment by milqnomadu** [comment]
@EmilDohne thank you Emil. I've sent you an email a couple of days ago, don't know if you saw it though.

**Comment by EmilDohne** [comment]
Hey @milqnomadu I responded now it managed to land in my spam, thanks a lot!

**Comment by EmilDohne** [comment]
Hey there @milqnomadu, it's been quite a while since I was fairly busy with another project but I have picked this up again and am currently working on implementing Smart Objects as the first set of Layers that come from the recent changes :)

**Comment by milqnomadu** [comment]
Hey @EmilDohne happy to hear about this! The key feature for smart objects would be to be able to change the image inside of it while maintaing the smart object’s warp, distort, opacity and position property!

**Comment by EmilDohne** [comment]
Hey @milqnomadu from what I saw so far the smart object stores all of this information as well as the raw file bytes in the file itself. It may be possible that photoshop applies the warp on load rather than storing it in the pixels directly which would make this a whole lot easier.

**Comment by milqnomadu** [comment]
Hi @EmilDohne, for sure. Do you have any idea on how would this be possible or how does the “on load” rather than “stored” warp affect the implementation?

**Comment by EmilDohne** [comment]
If Photoshop recreates the warp on load just modifying the data should be enough to transfer a warp. If instead it warps the actual pixels I'd have to write a warp implementation which would come later with the rendering pipeline :)

**Comment by EmilDohne** [comment]
This is mostly for my information/safekeeping but maybe you will find it interesting too. Essentially photoshop stores (with the default warp) a 4x4 polygon grid of the warp information. I initially assumed that this grid defines the actual polygon grid that then gets smoothed. Instead it defines the actual handles with which the warp happens. 

So for each quad there is the control point (corner) and two handles with the central vertex just being there to make it a quad.

The actual deformation is then implemented using a bezier surface defined by these points

## The quad mesh provided by the photoshop warp descriptor
![image](https://github.com/user-attachments/assets/4de11198-d99b-4ece-b393-b5e3b768b6f3)
> This grid was generated with some rendering functionality I added 

## The generated bezier surface, notice the handles overlap the points shown in the previous example
![image](https://github.com/user-attachments/assets/9afde44a-8a38-4dd6-9f65-c903233448ce)

**Comment by EmilDohne** [comment]
Alright I've managed to Implement a ` struct BezierSurface` which can be used to sample the point

![image](https://github.com/user-attachments/assets/2dad0ee4-6f91-430b-b42f-881843bcd2ca)

**Comment by EmilDohne** [comment]
I've added another visualization mode and you can see our rendered result matches what photoshop has 1 to 1:

# Photoshop
![image](https://github.com/user-attachments/assets/d5d0b8f5-addf-45e5-af17-be7a1ae82e6c)

# Our warp rendering
![image](https://github.com/user-attachments/assets/a2b906f2-19fc-4fb8-a99c-0cabd6e95475)

**Comment by milqnomadu** [comment]
@EmilDohne Awesome, very interesting indeed! Thank you so soo much!! So is it now possible to get and apply the warp with PhotoshopAPI using  struct BezierSurface?

**Comment by EmilDohne** [comment]
@milqnomadu not quite yet, there's still some work to be done to actually apply the warps but the framework is in place. There is a `SmartObjectWarp` struct which would actually store and roundtrip the warp information.

I still have to think of a nice way to author these but roundtripping will be supported for sure

**Comment by EmilDohne** [comment]
The actual warping is in place now too, this just a simple UV grid being deformed for now. This is still very slow so I'll have to work on that 

![image](https://github.com/user-attachments/assets/061e6667-915b-493d-bd08-9387a195343c)

**Comment by EmilDohne** [comment]
I've now added support for all types of warps (i.e. if you warp and you add your own subdivision rather than using the 4x4 grid)

![warp_applied_uv_tex](https://github.com/user-attachments/assets/8fd47635-9130-43c8-9316-c82ea3c6261c)

Now what is left to do is to add support for perspective and skew warps, although I suspect this isnt too difficult.

The performance is also now at a very reasonable level. I.e. the warp now takes only around a quarter of a second on an image of 7500x4300 pixels

**Comment by milqnomadu** [comment]
@EmilDohne Incredible work Emil, you're so talented! Thank you so much for everything you did, I tried helping you the best I could. I'll wait anxiously for perspective swell ahah. By the way I tried checking PhotoshopAPI's docs but I couldn't find the guidelines for this new warp feature, do you have any advice on how to implement it by getting the warp from a SO and applying it changing the image inside the SO recursively via Python?

**Comment by EmilDohne** [comment]
The warp isnt merged into the master branch yet, I'll likely need another bit to actually get it properly working since theres a lot of edge cases to account for but I'll try my best :)

**Comment by EmilDohne** [comment]
I've added support for all types of warps that arent smart filters now (smart filters will have to wait a bit). I can now get into applying it and creating a nice interface for users to apply, replace etc.

Here you can see a more advanced case of having both a skew and a warp applied at the same time as these are stored differently 

# Warp:
![image](https://github.com/user-attachments/assets/79f461f9-d349-4b85-80dc-25f72ab07150)


# Rendered checkerboard from the api
![warp_grid_u](https://github.com/user-attachments/assets/4ad062ad-861b-4895-bfce-a3b741b9d660)

I'll let you know once i've merged this

**Comment by milqnomadu** [comment]
@EmilDohne amazing! I wanted to ask you whether once it is ready to use if you could please update me. Thanks emi

**Comment by bonoem** [comment]
love the

**Comment by EmilDohne** [comment]
Hey @milqnomadu, sorry for the late reply but sure it's possible! Although it is a bit involved. For the example I will make the following assumptions:

- The layer has a regular photoshop warp applied to it
- We will not handle rescaling of the layer (the warp will be applied in the dimensions of the original image)
- The layer is not skewed, rotated etc. on top of the warp

I have implemented more or less what I have on the cpp side to pure python with these assumptions (makes it easier to recreate).
https://github.com/EmilDohne/psd_tools_apply_warp

You can check out the `main.py` file and run it, its entirely self contained (with the dependencies mentioned in requirements.txt). To use a different file for extracting the warp information just change out the psd to load and to apply it to a different image choose a different path here

```py
buffer.save(os.path.join(os.path.dirname(__file__), "out", "output_warped.jpg"))
```

This code does a whole bunch of things and might not be perfect (feel free to tinker around with it!) but it essentially takes the photoshop file, extracts the warp information and applies it to a new image.

So in short:

- load this photoshop warp: 
![image](https://github.com/user-attachments/assets/db3a4621-0f91-409b-ae29-6f4a9e2fd18d)

- extract the points into a mesh representation 
![WarpVariousNdims_bezier](https://github.com/user-attachments/assets/362ab5ab-581a-4c38-9e99-fe3edae4aaf0)

- sample this mesh as if it was a bezier and subdivide it
![WarpVariousNdims_mesh](https://github.com/user-attachments/assets/aac1a951-43e4-437b-8b37-8c492ef50c95)

- load another image (in this case its the same but can be a different one)
![image](https://github.com/user-attachments/assets/9c96e22d-57f2-43c5-93a3-84f18997a6c4)

- reapply the mesh warp and save it out as jpg
![output_warped](https://github.com/user-attachments/assets/2d41b44e-4e49-4a39-8bda-c8e26b8c3b26)


## Closing words

As I said, this is by no means perfect and skips over transformations, homography etc. but for the file you sent me a while ago this should suffice. Performance is also not great as its pure python but better than nothing. Doing this for an output resolution of 520*282 pixels took 5 seconds on my machine. (extracting that, 1040 px would be 20 seconds, 2080 px would be 80 seconds etc.)

Hopefully it gives you a decent idea of what is happening under the hood. Sorry for the lack of comments and readme. I'll try and get around to it but if you need to revisit the concepts I have fairly comprehensive documentation on the current state of the cpp implementation this is based on:

- [Warp class storing all transformations](https://github.com/EmilDohne/PhotoshopAPI/blob/57-add-support-for-adjustment-layers/PhotoshopAPI/src/Core/Warp/SmartObjectWarp.h)
- [ImageBuffer representation with rescaling etc.](https://github.com/EmilDohne/PhotoshopAPI/blob/57-add-support-for-adjustment-layers/PhotoshopAPI/src/Core/Render/ImageBuffer.h)
- [Bezier Surface](https://github.com/EmilDohne/PhotoshopAPI/blob/57-add-support-for-adjustment-layers/PhotoshopAPI/src/Core/Geometry/BezierSurface.h)
- [Mesh with Octree acceleration structure](https://github.com/EmilDohne/PhotoshopAPI/blob/57-add-support-for-adjustment-layers/PhotoshopAPI/src/Core/Geometry/Mesh.h)

**Comment by EmilDohne** [comment]
I realize now that I may have jumped the gun there and what you were asking wasnt photoshop specific but rather how you could do this in your own package? 

Even if I may have been overzealous there the code in the repository can be generally applied outside of photoshop too, the only thing really photoshop specific is the main.py which reads out that data

**Comment by EmilDohne** [comment]
I've still got some kinks to iron out related to transforms and writing of these files is currently unimplemented but I thought I'd give an update on what the api looks like roughly at this stage (on the C++ side)

```cpp
auto main() -> int
{
  using namespace NAMESPACE_PSAPI;
  
  LayeredFile<bpp8_t> file = LayeredFile<bpp8_t>::read("warptest_single.psd");
  
  auto layer_ptr = find_layer_as<bpp8_t, SmartObjectLayer>("WarpQuilt", file);

  auto orig_image_data = layer_ptr->original_image_data();
  // Handles applying the warp and transforms, no need to do it manually
  auto image_data = layer_ptr->get_image_data();
  
  write_to_disk(orig_image_data, "original.jpg", layer_ptr->original_width(), layer_ptr->original_height());
  write_to_disk(image_data, "warped.jpg", layer_ptr->width(), layer_ptr->height());
}
```

**Comment by milqnomadu** [comment]
Yes, in the meantime i wanted to try and recreate this GUI in python https://www.reddit.com/r/computervision/comments/1dtmln4/how_to_replicate_photoshops_transform_warp_tool/ so i can either set the warp myself or it just retrieves the warp from the smart object in the psd file i give it and it will then apply it to another image. I've been trying since forever but can't figure it out or find anything related anywhere.

**Comment by EmilDohne** [comment]
@milqnomadu you should be able to lift the code from the repo posted above and it will do what you want but wont do it realtime. Unfortunately if you want something efficient you'll likely have to wait for the v0.6.0 release :(

**Comment by zzq0324** [comment]
Great work, awesome. I really need these technical points, but my ability is limited. I am glad you can share it.

**Comment by milqnomadu** [comment]
Thank you so much Emil! (I missed the part where you created the repo and only noticed now, sorry!)

**Comment by kratosvn** [comment]
amazing @EmilDohne <3

**Comment by EmilDohne** [comment]
This is just a text based update, but I've just gotten file writes with smart object layers to work with the warp staying live (it still complains when replacing the image). One step closer I suppose :D

**Comment by EmilDohne** [comment]
Another update, now replacing images or creating layers from scratch works as well as transferring warps from one layer to another. The warps then stay live in photoshop

```cpp
LayeredFile<bpp8_t> file = LayeredFile<bpp8_t>::read("warp.psd");

// Relink the file to a different layer
auto layer_ptr = find_layer_as<bpp8_t, SmartObjectLayer>("WarpQuilt", file);
layer_ptr->replace("uv_grid.jpg");

// Create a new smart obbject based on a file and copy over the warp from one layer to another
auto params = Layer<bpp8_t>::Params{};
params.name = "UVGrid";
auto layer_new = std::make_shared<SmartObjectLayer<bpp8_t>>(file, params, "uv_grid.jpg");
layer_new->warp(layer_ptr->warp());

file.add_layer(layer_new);
LayeredFile<bpp8_t>::write(std::move(file), "warp_out.psd");
```

![image](https://github.com/user-attachments/assets/240dceb3-91db-46a4-b1ba-117579c32728)


I've also gotten the tests to run again so tomorrow I will expand on them and then soon work on the python bindings for smart objects

**Comment by EmilDohne** [comment]
Making some more progress, I've now also added an example on the cpp side. Tomorrow I'll start work on the python bindings although I doubt I'll be done with it before the new year. Here's a teaser of what can already be done on the cpp side though!

```cpp
int main()
{
	using namespace NAMESPACE_PSAPI;

	// Reading the file works as normal, as well as accessing the layers.
	LayeredFile<bpp8_t> file = LayeredFile<bpp8_t>::read("SmartObject.psd");
	auto layer_ptr = find_layer_as<bpp8_t, SmartObjectLayer>("warped", file);

	// If we want to add an additional smart object layer we can do this just like you would with a normal layer.
	// The only exception is that we have to provide a path for the image file we wish to link. We can additionally provide a warp
	// object which will reapply the given warp to the new layer. This is convenient if you wish to create a copy of a layer without
	// removing the other layer
	auto params = Layer<bpp8_t>::Params{};
	params.name = "uv_grid";
	auto layer_new = std::make_shared<SmartObjectLayer<bpp8_t>>(file, params, "uv_grid.jpg", layer_ptr->warp(), LinkedLayerType::external);

	// Smart objects also have more freedom in their transformations which we expose in the API:
	layer_new->rotate(45);	// Rotate by 45 degrees
	layer_new->scale(.65f);	// Scale by a factor of .65
	layer_new->move(Geometry::Point2D<double>(50.0f, 0.0f));	// Move to the right by 50

	//// We can additionally also apply a perspective transformation using a 3x3 projective matrix.
	//// This could also be used to skew or to directly apply all the above operations. This matrix e.g.
	//// applies a perspective transform where the vanishing point is at 2048, 0 (keeping in mind that 0 
	//// in this case is at the top of the canvas). 
	//// If you wish to know more about how these 3x3 transformation matrices work this is a great video/channel:
	//// https://www.youtube.com/watch?v=B8kMB6Hv2eI.
	Eigen::Matrix3d transformation;
	transformation << 1, 0, 0, 
                          0, 1, 0,
                          1.0f / 2048, 0, 1;
	layer_new->transform(transformation);

	// If you want to be a bit more descriptive of how these matrices are to be built you can create one using a source
	// and destination quad like this. Keep in mind that these have to be in the coordinate space of the image itself:
	std::array<Geometry::Point2D<double>, 4> source_transform = Geometry::create_quad<double>(2048, 2048);
	std::array<Geometry::Point2D<double>, 4> destination_transform = Geometry::create_quad<double>(2048, 2048);
	// Squash together the top of the image
	destination_transform[0].x = 512;
	destination_transform[1].x = 1536;

	// Create a 3x3 homography mapping the source_transform quad to the destination_transform quad.
	auto homography = Geometry::Operations::create_homography_matrix(source_transform, destination_transform);
	layer_new->transform(homography);

	// If now we wanted to clear the transformation and/or warp that could be done as well. This will now just
	// be a smart object layer with width and height of 2048 (like the original image data).
	layer_new->reset_transform();
	layer_new->reset_warp();

	// For modifying the warp directly (bezier) we can push the individual points, although this is a bit more
	// advanced and we don't currently have a very convenient interface for it. This code right here 
	// pushes the top left corner to the 50, 50 position which will create a slightly rounded corner
	auto& warp = layer_new->warp();
	warp.point(0, 0).x = 50.0f;
	warp.point(0, 0).y = 50.0f;

	// adding these works just as with any other layers, writing also works as usual
	file.add_layer(layer_new);
	LayeredFile<bpp8_t>::write(std::move(file), "SmartObject_out.psb");
}
```

**Comment by bonoem** [comment]
@EmilDohne have you had any issues?

**Comment by EmilDohne** [comment]
Hey @bonoem, no issues per se. Progress is going good :). I've got the python bindings in and am currently working on ironing out kinks related to transforms and making sure that the warps produced by the photoshopapi are close enough to the ones produced by photoshop itself.

As usual the biggest limiting factor is unfortunately time

**Comment by bonoem** [comment]
> Hey @bonoem, no issues per se. Progress is going good :). I've got the python bindings in and am currently working on ironing out kinks related to transforms and making sure that the warps produced by the photoshopapi are close enough to the ones produced by photoshop itself.
> 
> As usual the biggest limiting factor is unfortunately time

Awesome!

**Comment by EmilDohne** [comment]
Alright so I have a bit of an update, I've worked primarily on getting both anti-aliased edges and making the transformations correct for all cases and I've managed to get the mean error between layers exported from the PhotoshopAPI and Photoshop down to `<0.004` (0.4%). 
Here i've got a sample image of one of the test cases showing off both a warp and a perspective transform, this image was written out using the PhotoshopAPI and except for the edges being slightly different is near-identical to the original from photoshop

![Image](https://github.com/user-attachments/assets/dbcb7338-b179-4eb8-826a-79b02f491827)

For reference, this is what the two images diffed look like with the only perceptible difference being the edges likely due to using a different anti-aliasing approach (I use supersampling)

![Image](https://github.com/user-attachments/assets/980e262c-2942-40d2-84eb-becd13c8f769)

I'll now start work on finalizing the python bindings, test cases and docs and then should be good to merge this :)

**Comment by EmilDohne** [comment]
Unfortunately due to some pybind related issues I'm currently in the process of rewriting the image data mixins a bit so it'll take a bit longer

**Comment by kratosvn** [comment]
> Unfortunately due to some pybind related issues I'm currently in the process of rewriting the image data mixins a bit so it'll take a bit longer

i'm waiting for good news

**Comment by EmilDohne** [comment]
I'm more or less done with this feature now. I'll put it in a MR sometime early next week and merge it. I'll then post it as a pre-release (that you can still download from pip) so you guys can test it for a week or two before it goes live :)

**Comment by EmilDohne** [comment]
This is merged and available as version 0.6.0.

Please do give it a try and let me know if there's  any issues with it! I'll close this ticket but feel free to create new ones if something pops up!

**Comment by kratosvn** [comment]
Thank you so much, @EmilDohne  I really appreciate your effort in completing and releasing version 0.6.0. I'll give it a try and let you know if anything comes up. Truly grateful for your dedicated support!

**Comment by milqnomadu** [comment]
@EmilDohne any updates?

**Comment by EmilDohne** [comment]
Hey @milqnomadu , this has been in the PhotoshopAPI for a couple months now, or do you mean something else?

**Comment by milqnomadu** [comment]
@EmilDohne Yes. I meant regarding the inability to apply properties like warp or perspective on of the smart object to the image, making it not so useful

**Comment by bonoem** [comment]
@EmilDohne how is it going?

**Comment by EmilDohne** [comment]
Hey @bonoem , if you mean regarding #142 that should be fixed and addressed by now! Or is there something else?

**Comment by bonoem** [comment]
@EmilDohne Hey Emil, oh wonderful. So now it's possibile to insert a new image inside a smart objects and have the file exported with the warp of the SO applied correctly? Are the doc also updated to learn how to create this code?

**Comment by EmilDohne** [comment]
The example repo was only ever meant to bridge the gap between having something and releasing it into the photoshopapi. Now it's deprecated and doesn't really serve any purpose except for historical :)

As for your second question, sure that's possible! Here's some example code (semi-pseudo, haven't tested) for doing exactly that:

```py

filepath_psd = "my_psd_file_path.psd"
filepath_png = "my_png_file_path.png"

psd_file = psapi.LayeredFile.read(filepath_psd)

#either:
smart_object = psd_file["layer_name"]
# or:
smart_object = None
for layer in psd_file.layers:
   if isinstance(layer, (psapi.SmartObjectLayer_8bit, psapi.SmartObjectLayer_16bit, psapi.SmartObjectLayer_32bit)):
      smart_object = layer
      break


# easiest way to do this is to replace the smart object with the image and then get the image data
smart_object.replace(filepath_png)
image_data: Dict[int, np.ndarray] = smart_object.get_image_data

# Now use whatever library you want to write out the channels to webp/jpeg/png
```

**Comment by bonoem** [comment]
Oh ok. Cause I can't manage to code what I told you ahahha
